kornia/geometry/transform/homography_warper.py

```diff
@@ -164,6 +168,109 @@ def homography_warp3d(
+                 loss_fn = F.l1_loss,
+                 pyr_levels = 5,
+                 lr: float = 1e-3,
+                 n_iter: int = 100):
+        super().__init__()
+        self.model_type = model_type
+        if model_type == "H":
+            self.warper = HomographyWarper
+            self.model = Homography()
+        else:
+            raise ValueError(f"{model_type} is not supported. Try 'H'")
+        self.pyr_levels = pyr_levels
+        self.optimizer = optimizer
+        self.lr = lr
+        self.loss_fn = loss_fn
+        self.n_iter = n_iter
+
+    def get_single_level_loss(self,
+                              img_src: torch.Tensor,
+                              img_dst: torch.Tensor,
+                              transform_model: torch.nn.Module) -> torch.Tensor:
+        # ToDo: Make possible registration of images of different shape
+        assert len(img_src.shape) == len(img_dst.shape), (img_src.shape, img_dst.shape)
+        _height, _width = img_src.shape[-2:]
+        warper = self.warper(_height, _width)
+        img_src_to_dst = warper(img_src, transform_model)
+        # compute and mask loss
+        loss = self.loss_fn(img_src_to_dst, img_dst, reduction='none') # 1x3xHxW
+        ones = warper(torch.ones_like(img_src), transform_model)
+        loss = loss.masked_select(ones > 0.9).mean()
+        return loss
+
+    def reset_model(self):
+        self.model.reset_parameters()
+
+    def register(self,
+                 src_img: torch.Tensor,
+                 dst_img: torch.Tensor,
+                 verbose: bool = False,
+                 device=torch.device('cpu')) -> torch.Tensor:
+        '''Example:
+        >>>> registrator = ImageRegistration('H', lr=1e-3, n_iter=50, pyr_levels=5,
+                               loss_fn = F.l1_loss)
+        >>>> homo = registrator.register(img_src, img_dst, True)
+        '''
+        # compute the gaussian pyramids
+        self.to(device)
+        self.reset_model()
+        # ToDo: better parameter passing to optimizer
+        opt = self.optimizer(self.model.parameters(), lr=self.lr)
+        img_src_pyr = build_pyramid(src_img.to(device), self.pyr_levels)
+        img_dst_pyr = build_pyramid(dst_img.to(device), self.pyr_levels)
+        assert len(img_dst_pyr) == len(img_src_pyr)
+        for img_src_level, img_dst_level  in zip(img_src_pyr, img_dst_pyr):
+            for i in range(self.n_iter):
+                # compute gradient and update optimizer parameters
+                opt.zero_grad()
+                loss = self.get_single_level_loss(img_src_level,
+                                                  img_dst_level,
+                                                  self.model())
+                loss.backward()
+                if verbose and (i % 10  == 0):
+                    print (f"Loss = {loss.item():.4f}, iter={i}")
+                opt.step()
+        return self.model
+
+    def warp_src_into_dst(self, src_img, dst_img):
```

Comment:
in theory K.geometry.warp_perspective should work too, right ? probably add as a TODO

Reply:
Here - yes, but I want to have it model-agnostic.

Reply:
true !